htmresearch/encoders/cio_encoder.py

```diff
@@ -361,3 +374,19 @@ def densifyPattern(self, bitmap):
     for i in bitmap:
       sparsePattern[i] = 1.0
     return sparsePattern
+
+
+  def reduceSparsity(self, encoding, maxSparsity):
+    """Reduce the sparsity of the encoding down to maxSparsity"""
+
+    desiredBits = maxSparsity*encoding["width"]*encoding["height"]
+    bitmap = encoding['fingerprint']['positions']
+
+    # Choose a random subsampling of the bits but seed the random number
+    # generator so we get consistent bitmaps
+    numpy.random.seed(bitmap.sum())
+    encoding['fingerprint']['positions'] = (
+      numpy.random.permutation(bitmap)[0:desiredBits] )
```

Comment:
Because line length was too long. I think this way of breaking it up is a bit clearer than something like:
```
    encoding['fingerprint'][
      'positions'] = numpy.random.permutation(bitmap)[0:desiredBits]
```